Document MaybeUninit bit validity

joshlf opened the pull request:
Partially addresses https://github.com/rust-lang/unsafe-code-guidelines/issues/555 by clarifying that it is sound to write any byte values (initialized or uninitialized) to any `MaybeUninit<T>` regardless of `T`.

r? @RalfJung

joshlf committed:
Document MaybeUninit bit validity

Co-authored-by: Ralf Jung <post@ralfj.de>
Edited-by: TC

joshlf commented on library/core/src/mem/maybe_uninit.rs:
````diff
@@ -252,6 +252,30 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// }
+/// ```
+///
+/// If `T` contains initialized bytes at byte offsets where `U` contains padding bytes, these
+/// may not be preserved in `MaybeUninit<U>`, and so `transmute(u)` may produce a `T` with
+/// uninitialized bytes in these positions. This is an active area of discussion, and this code
+/// may become sound in the future.
````
Per https://github.com/rust-lang/unsafe-code-guidelines/issues/555#issuecomment-2839258901, I've updated to the following text. Does that look good?

```rust
/// Note that, so long as every byte position which is initialized in `T` is also initialized
/// in `U`, then the preceding `identity` example *is* sound.
```

RalfJung replied:
As noted below I don't think the term "bytes initialized in `T`" makes a lot of sense. But by taking a reasonable guess at what you mean by this (non-padding byte), then yes I think that is sound.

joshlf commented on library/core/src/mem/maybe_uninit.rs:
````diff
@@ -252,6 +252,33 @@ use crate::{fmt, intrinsics, ptr, slice};
 /// }
 /// ```
+///
+/// # Validity
````
Moving [this discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/286#issuecomment-2839284938) here:

> > The `MaybeUninit` docs probably make sense for this. We now do have a [definition of "byte"](https://doc.rust-lang.org/nightly/reference/memory-model.html#bytes) in the reference that this can link to.
> 
> Okay, awesome. And what wording would you recommend? Would it be accurate to say something like the following?
> 
> > The value of a `[MaybeUninit<u8>; N]` may contain pointer provenance, and so `p: P -> [MaybeUninit<u8>; N] -> P` preserves the value of `p`, including provenance

@RalfJung would you like me to add language like this to this PR?

joshlf replied:
Updated to reference the [definition of a byte](https://doc.rust-lang.org/nightly/reference/memory-model.html#bytes).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14734800786/job/41358187411) [(plain)](https://github.com/rust-lang/rust/commit/21626ac072d2d15df29fc3ec94f84d3331675e02/checks/41358187411/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- library/core/src/mem/maybe_uninit.rs - mem::maybe_uninit::MaybeUninit (line 266) stdout ----
error[E0133]: call to unsafe function `std::intrinsics::transmute` is unsafe and requires unsafe block
 --> library/core/src/mem/maybe_uninit.rs:270:29
  |
7 |     let u: MaybeUninit<U> = transmute(t);
  |                             ^^^^^^^^^^^^ call to unsafe function
  |
  = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function `std::intrinsics::transmute` is unsafe and requires unsafe block
```

</details>

RalfJung commented:
Cc @rust-lang/opsem

RalfJung commented on library/core/src/mem/maybe_uninit.rs:
````diff
@@ -252,6 +252,44 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// }
+/// ```
+///
+/// If `T` contains initialized bytes at byte offsets where `U` contains padding bytes, these
+/// may not be preserved in `MaybeUninit<U>`, and so `transmute(u)` may produce a `T` with
+/// uninitialized bytes in these positions. This is an active area of discussion, and this code
+/// may become sound in the future.
````
I don't think it makes sense to say that a *type* "contains initialized bytes" at some offset. That's a property of a representation.

The typical term for representation bytes that are lost here is "padding". I don't think we have rigorously defined padding anywhere yet, but the term is sufficiently widely-used (and generally with a consistent meaning) that we may just be able to use it here?

joshlf replied:
IIUC, you're making two points:
- We should speak about a type's representation containing bytes, not about the type itself containing bytes
- In a representation, we should speak about padding bytes rather than uninitialized bytes

Is that right?

One thing that's probably worth distinguishing here is between values and layouts. In my mental model, an uninit byte is one of the possible values that a byte can have (e.g., it's the 257th value that can legally appear in a `MaybeUninit<u8>`). By contrast, padding is a property of a layout - namely, it's a sequence of bytes in a type's layout that happen to have the validity `[MaybeUninit<u8>; PADDING_LEN]`.

Based on this, maybe it's best to say:

> If byte offsets exists at which `T`'s representation does not permit uninitialized bytes but `U`'s representation does (e.g. due to padding), then the bytes in `T` at these offsets may not be preserved in `u`, and so `transmute(u)` may produce a `T` with uninitialized bytes at these offsets. This is an active area of discussion, and this code may become sound in the future.

RalfJung replied:
> Is that right?

No. I think both of the following concepts make sense:
- The representation of a particular value at a particular type contains uninitialized bytes
- A type contains padding bytes

But it makes less sense to talk about padding of a representation, or to talk about uninitialized bytes in a type.

So for this PR, the two key points (and they are separate points) are:
- If `U` has padding, those bytes may be reset to "uninitialized" as part of the round-trip. If those same bytes are *not* padding in `T`, this can therefore mean some of the information of the original `T` value is lost.
- If `T` does not permit uninitialized bytes on those positions, the round-trip is UB.

The second point is just a logical consequence of the first, it does not add any new information. Not sure if it is worth mentioning.

joshlf replied:
> * The representation of a particular value at a particular type contains uninitialized bytes.
> * A type contains padding bytes. (These are bytes which are always ignored by the representation relation.)

Does this imply that a *type* contains padding bytes, not a type's *representation*?

I'm thinking through the implications of what you said, and I think I understand something new that I didn't before, and I want to run it by you: In my existing mental model, a padding byte is a location in a type's layout such that every byte value at that location (including uninit) is valid (enums complicate this model, but I don't think that complication is relevant for this discussion - we can just stick to thinking about structs). The problem with this mental model is that, interpreted naively, it implies that *different* byte values in a padding byte could correspond to *different* logical values of the type. So e.g. in the type `#[repr(C)] struct T(u8, u16)`, `[0, 0, 0, 0]` and `[0, 1, 0, 0]` would correspond to different values of the type since we're treating the padding byte itself as part of the representation relation. Of course, that is not something we want.

IIUC, by contrast your model is that the representation relation simply doesn't include padding bytes at all. So it'd be more accurate to describe the representation of `T` as consisting of *three* bytes - at offsets 0, 2, and 3. Every representation of `T` has a "hole" at offset 1 which is not part of the representation. This ensures that there's a 1:1 mapping between logical values and representations. Is that right?

RalfJung replied:
> Does this imply that a type contains padding bytes, not a type's representation?

That's how I think about it. We can't tell which byte is a padding byte by looking at one representation -- it's a property of the type.

> In my existing mental model, a padding byte is a location in a type's layout such that every byte value at that location (including uninit) is valid

That would make the only byte of `MaybeUninit<u8>` a padding byte, so I don't think this is the right definition.
That's why I said above: a padding byte is a byte that does not matter for the representation. Slightly more formally: if `r` is some representation valid for type `T`, and `r'` is equal to `r` everywhere except for padding bytes, then `r` and `r'` represent the same value.

> So it'd be more accurate to describe the representation of T as consisting of three bytes 

The representation has 4 bytes. But only 3 of them actually affect the represented value (which is a tuple of two [mathematical] integers).

We seem to be using the term "representation" slightly different. For me, that's list a `List<Byte>` of appropriate length. You may be using that term to refer to what I call ["representation relation"](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#representation-relation)?

RalfJung commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,44 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// # Provenance
+///
+/// `MaybeUninit` values may contain [pointer provenance][provenance]. Concretely, for any
+/// pointer type, `P`, which contains provenance, transmuting `p: P` to
```
I would say "for any value `p: P` that contains provenance", or so -- "types that contain provenance" doesn't make much sense to me, and restricting this to pointer types seems unnecessary (this also applies to e.g. tuples and arrays containing pointers).

joshlf replied:
Good point; updated to use your language.

RalfJung commented:
@rustbot ready

ia0 commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,47 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// may not be preserved in `MaybeUninit<U>`, and so `transmute(u)` may produce a `T` with
+/// uninitialized bytes in these positions. This is an active area of discussion, and this code
+/// may become sound in the future.
+
+/// If byte offsets exists at which `T`'s representation does not permit uninitialized bytes but
+/// `U`'s representation does (e.g. due to padding), then the bytes in `T` at these offsets may
+/// not be preserved in `u`, and so `transmute(u)` may produce a `T` with uninitialized bytes at
+/// these offsets. This is an active area of discussion, and this code may become sound in the future.
```
Doesn't this repeat the above? I guess it's a left-over because there's also an empty line instead of `///`.

joshlf replied:
Oops, good catch. Removed.

ia0 commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,47 @@ use crate::{fmt, intrinsics, ptr, slice};
+///
+/// `MaybeUninit` values may contain [pointer provenance][provenance]. Concretely, for any
+/// value, `p: P`, which contains provenance, transmuting `p` to `MaybeUninit<[u8; size_of::<P>]>`
+/// and then back to `P` will produce a value identical to `p`, including provenance.
```
Doesn't this either contradict the above or assume that `P` does not have padding bytes? (`P` could be a struct containing pointers for example)

RalfJung commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,47 @@ use crate::{fmt, intrinsics, ptr, slice};
+///
+/// Note that, so long as no such byte offsets exist, then the preceding `identity` example *is* sound.
+///
+/// # Provenance
```
Should we really discuss provenance separately?

The way I'd structure this is I'd first explain that `MaybeUninit` has no validity invariant, ergo it can hold arbitrary data, ergo `T `→ `[MaybeUninit<u8>; size_of::<T>()]` → `T` is a valid round-trip, including if `T` has padding or provenance. And then we can talk about the other round-trip which is not legal.

joshlf replied:
Updated to make it clear that this is just an implication which I'm calling out. Lmk if that looks better?

RalfJung commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,47 @@ use crate::{fmt, intrinsics, ptr, slice};
+///
+/// Note that, so long as no such byte offsets exist, then the preceding `identity` example *is* sound.
+///
+/// # Provenance
```
FWIW, this guarantee is *not* currently reflected in Alive's semantics for the LLVM IR we generate. We compile `MaybeUninit<u8>` to `i8` and Alive says that loading `i8` from memory with provenance is UB.

But this is more of a deeper reflection of the fact that the semantics we want for `MaybeUninit` are fundamentally impossible to express in LLVM right now. We can either wait until LLVM has a way to express this, or we just hope that this won't go wrong in practice... @nikic I wonder if you have any thoughts on this.

RalfJung commented:
@rustbot author

karolzwolak commented:
Thanks for your contribution @joshlf from wg-triage.
Could you address the comments above?

joshlf commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,47 @@ use crate::{fmt, intrinsics, ptr, slice};
+///
+/// A `MaybeUninit<T>` has no validity requirement – any sequence of bytes of the appropriate length,
+/// initialized to any value or uninitialized, are a valid value of `MaybeUninit<T>`. Equivalently,
+/// it is always sound to perform `transmute::<[MaybeUninit<u8>; size_of::<T>()], MaybeUninit<T>>(...)`.
```
Re-reading it, I agree that it's unnecessary. Removed.

joshlf commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,47 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// initialized to any value or uninitialized, are a valid value of `MaybeUninit<T>`. Equivalently,
+/// it is always sound to perform `transmute::<[MaybeUninit<u8>; size_of::<T>()], MaybeUninit<T>>(...)`.
+///
+/// Note that "round-tripping" via `MaybeUninit` does not always result in the original value.
```
Done.

joshlf commented:
> > Thanks for your contribution @joshlf from wg-triage. Could you address the comments above?
> 
> I likely won't have time to move this forward until mid-September or October, but I'll follow up at that point.

Nvm, found some time 🙂 

I've responded to various comment threads.

karolzwolak commented:
Awesome, you should also rebase your changes onto master, and do `@rustbot ready` when you're ready for review.

joshlf commented:
@rustbot ready

RalfJung commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,41 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// However, so long as no such byte offsets exist, then the preceding `identity` example *is* sound.
+/// In particular, since `[u8; N]` has no padding bytes, transmuting `t` to `MaybeUninit<[u8; size_of::<T>]>`
+/// and back will always produce the original value `t` again. This is true even if `t` contains [provenance]:
+/// the resulting value will have the same provenance as the original `t`.
```
Seems tricky to do that in a non-confusing way...

Let's see what some of the other folks say. TBH my inclination is to go with
> We can also argue that the transmute itself perfectly preserves provenance, the reborrow happens in the surrounding code.

since that's what actually happens in MiniRust.

RalfJung replied:
Here's the best I can come up with so far. But I feel like this may actually be easier to write for someone who's less deeply entrenched in these discussions. ;)
> Note: if `t` contains a reference, then there may be implicit reborrows of the reference any time it is copied, which may alter its provenance. In that case, the value returned by `identity` is not exactly the same as its argument. However, even in this case, it remains true that `identity` behaves the same as a function that just returns `t` immediately.

joshlf replied:
Okay here's my version of this (which is in the PR now):

> Note a potential footgun: if `t` contains a reference, then there may be implicit reborrows of the reference any time it is copied, which may alter its provenance. In that case, while the value returned by `identity` is exactly the same as its argument, that value may immediately be reborrowed upon return, altering its provenance. This may make this call to `identity` behave as though it does *not* exactly preserve provenance.

RalfJung replied:
> while the value returned by identity is exactly the same as its argument

That part isn't given (there can be reborrows inside `identity`).

In fact my comment explicitly stated "the value returned by identity may not be exactly the same as its argument" so not sure how you got from there to your version.

RalfJung replied:
That works for me :)

RalfJung commented on library/core/src/mem/maybe_uninit.rs:
````diff
@@ -252,6 +252,41 @@ use crate::{fmt, intrinsics, ptr, slice};
+/// ```
+///
+/// If the representation of `t` contains initialized bytes at byte offsets where `U` contains padding bytes, these
+/// may not be preserved in `MaybeUninit<U>`. Interpreting the representation of `u` at type `T` again (i.e., `transmute(u)` above) may thus
````
"interpret u as type T" sound wrong. We don't interpret u as a type, it's not a type. We use the type to tell us how to interpret u.

But I am not sure if there is widely-used standard terminology here, "interpret x at type y" just sounded most natural to me.

RalfJung replied:
That also works for me. I'd rather say "Transmuting `u` back to `T`" (to preserve the "again" in the current version).

joshlf replied:
Done!

traviscross committed:
Revise `MaybeUninit` validity documentation

Let's rewrite this for better clarity.  In particular, let's document
our language guarantees upfront and in positive form.  We'll then list
the caveats and the non-guarantees after.

RalfJung reviewed (commented):
Sounds great! Just minor nits.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/140463/ce4beebecb77821734079cff47d8af08f9f27f11..acade4f9797b978d393760d41626c67725baea05)*

joshtriplett commented on library/core/src/mem/maybe_uninit.rs:
```diff
@@ -252,6 +252,89 @@ use crate::{fmt, intrinsics, ptr, slice};
+///
+/// Moving or copying a value of type `MaybeUninit<T>` (i.e., performing a
+/// "typed copy") will exactly preserve the contents of all non-padding bytes of
+/// type `T` in the value's representation including the [provenance] of those
```
That does not seem at all clear; I think it would be both clearer and easier to read with the comma.
